README.md
# nanoboard
nanoboard client + sources in a single zip: https://github.com/nanoboard/nanoboard/releases

Внимание! Статья устарела и нуждается в обновлении

За время существования статьи спецификации поменялись и дополнились такими элементами как капча и Proof Of Work...
## Причина создания

 - Модераторский произвол
 - Мечта о неубиваемой борде

## Ценности Наноборды

 - простота формата нанопостов: хэш + текст сообщения = хэш (на который могут ссылаться другие нанопосты)
 - отсутствие владельца, прав, уровней доступа, невозможность централизованной модерации или отключения наноборды
 - отсутствие стандартизированного протокола распространения, в самой простой форме нанопост можно скопировать и вставить как текст
 - паразитирование на существующих ресурсах в интернете и скрытом интернете вместо организации p2p-соединений и/или серверов (низкий уровень детектирования + эксплуатация чужих мощностей)
 - поощряется помощь участникам и наноборде в виде ретрансляции чужих нанопостов (кроме огромных)
 - клиенты для просмотра собирают все нанопосты воедино и формируют цепочки - так сохраняется логическая связь между нанопостами и отпадает необходимость во временных метках (подверженных фальсификации)
 - клиенты для просмотра должны давать участнику возможность легко фильтровать нежелательный мусор, чтобы не видеть его и не распространять
 - клиенты для просмотра облегчают участнику распространение своих и чужих нанопостов (кроме мусора)
 - клиенты для просмотра ограничивают распространение слишком огромных чужих нанопостов, чтобы пресекать злоупотребление "трафиком" наноборды

## Спецификации

Каждый нанопост обязательно ссылается на другой нанопост поредством хэш-кода.

Хэш код 16-байтовый и выглядит как 32-символьный кусочек текста в шестандцатиричном текстовом виде.

Рассмотрим пример на JavaScript.

Нам понадобится SHA-256, взять можно например отсюда: http://www.movable-type.co.uk/scripts/sha256.html

или отсюда: https://username1565.github.io/sha256/sha256_utf8.html

```
var post = "00000000000000000000000000000000{Welcome to Nanoboard}"
var hash = Sha256.hash(post).substring(0,32) // берём хэш f682830a470200d738d32c69e6c2b8a43b44435f48ba000cfb5203b22e9bf026 в hex и отбрасываем всё кроме первых 32 символов
console.log(hash) // результат: f682830a470200d738d32c69e6c2b8a4
```
Хэш считается с байтов utf-8 кодировки. В nboard этот пост и его хэш используются как "корневые" но это не обязательно.

Теперь попробуем "ответить" на корневой пост.

```
var answer = hash + "Проверка ответа."
console.log(answer) // результат: f682830a470200d738d32c69e6c2b8a4Проверка ответа.
```

Ответим снова.

```
function getHash(content) { return Sha256.hash(content).substring(0,32); }
var answer2 = getHash(answer) + "Второй уровень."
console.log(answer2) // результат: 77bb9e5d1a64facc551210862e2c6536Второй уровень.
```
Вот так выглядят нанопосты, естественно, внутри можно ставить переносы строк и любые utf-8 символы.

Теперь осталось только распространить их. По хэшам все связывается в цепочки. Сложно сказать что является тредом среди нанопостов.

Определенно если на нанопост существует несколько ответов то с него можно начинать просмотр "треда".

Начать разворачивание можно с любого уровня, это уже на плечах реализаций просмотрщиков Наноборды.

В nboard можно разворачивать любые нанопосты кроме корневого, а можно просто просматривать только сам нанопост

и его ответы (следующий уровень).
_______
Наноборда берет начало с Анонимных Имидж Борд, Имидж означает наличие картинок (image).

Хотя существуют чисто текстовые борды, наноборда такой не является, потому как внутрь можно вставлять base64-изображения.

В nboard это реализовано так: если в нанопосте встречается конструкция вида [img=.....]

(где вместо точек - base64 строчка), то всё это дело превращается в изображение.

Кроме того nboard поддерживает разметку:

**[b]жирный[/b]**, *[i]курсив[/i]*, ~~[s]перечеркнутый[/s]~~, [sp]спойлер[/sp], [u]подчёркнутый[/u], [g]зелёный[/g].

Также, возможно распространять в виде base64 - любые файлы (размером до 65535 байт).

В основном, это zip-архивы, [но могут быть и любые другие файлы](https://github.com/nanoboard/nanoboard/issues/6).

Файлы, могут быть прекриплены, как JPEG-картинка, без сжатия, тогда, в нанопосте, используется:

 - либо старая конструкция [xmg=.....], где вместо точек - base64 несжатого файла,

   В этом случае, когда скрипт наноборды находит тег xmg, он автоматически добавляет ссылку на закачку zip-архива,

   превращая её в dataURL.

 - либо же используется bb-код file, с именем и типом файла - как у карасика.
 
   Тогда, при парсинге, имя файла попадает в ссылку, в атрибут download, а тип файла - в dataURL,
   
   и бинарный файл - можно закачать, по ссылке.

## PNG-контейнеры

На АИБ обычно можно свободно постить PNG-картинки. А в PNG-картинку можно вшить данные.

В nboard используется подход (и, собственно, код) из этой статьи: http://blog.andersen.im/2014/11/hiding-your-bits-in-the-bytes/ 

Вкратце суть в том что информация размещается в нижних битах компонентов R, G и B каждого пикселя

и это совершенно незаметно на глаз.

Разница между 24битным цветом и 21битным цветом с дизерингом (информациионный шум), по сути - минимальна.

Размер картинки после этого обычно увеличивается но не разительно.

Информация хранится именно в пикселях, что позволяет скопировать картинку в буфер обмена,

вставить в графический редактор, сохранить, скажем, в BMP, а потом пересохранить в PNG - вшитая информация не пропадёт.

Итак nboard берет некоторое количество свежих (по дате поступления на компьютер пользователя) постов,

а также некоторое количество случайно выбранных постов за всё время.

При этом избегаются скрытые пользователем посты, а крупные посты отправляются только если являются свежими.

В [наноборде карасика](https://github.com/Karasiq/nanoboard) также имеется кнопка "В очередь" и "Из очереди",

позволяющая тонко настроить список очереди нанопостов, для упаковки их оттуда - в контейнер.

Случайные посты, можно также отключить там.
_______
Но это там...

Здесь же, упаковка массива постов происходит следующим образом:

```
    public static byte[] Pack(NanoPost[] posts)
    {
        List<byte> bytes = new List<byte>();
        // количество постов (от 000000 до ffffff) (строчка в utf-8 байтах):
        bytes.AddRange(Encoding.UTF8.GetBytes(posts.Length.ToString("x6"))); 

        foreach (var p in posts)
        {
            // количество символов (не байтов!) в нанопосте (включая родительский хэш):
            var len = p.SerializedString().Length; 
            // тот же формат: от 000000 до ffffff (строчка в utf-8 байтах):
            bytes.AddRange(Encoding.UTF8.GetBytes(len.ToString("x6"))); 
        }

        foreach (var p in posts)
        {
            // utf-8 байты каждого поста по очереди без разделителей:
            bytes.AddRange(p.SerializedBytes()); 
        }

        // сжимаем результат gzip-ом
        return GZipUtil.Compress(bytes.ToArray());
    }
```

Перед упаковкой в контейнер (после сжатия gzip-ом) применяется шифрование salsa20 по некоторому ключу.

Для Наноборды пока есть один общий ключ, но дальше можно отделяться в отдельные зашифрованные Наноборды

(что нежелательно но иногда может быть необходимо):

```
        // Параметры инициализации salsa20 для заданного ключа (key):
        byte[] initKey = SHA256.Create().ComputeHash(Encoding.UTF8.GetBytes(key));
        byte[] initVec = SHA256.Create().ComputeHash(Encoding.UTF8.GetBytes(key).Reverse().ToArray());
        initKey = initKey.Crop(32);
        initVec = initVec.Crop(8);
```
_______
На этом пока что всё.

Напомню, что PNG-контейнеры - [не единственный возможный способ](https://username1565.github.io/js-jpeg-steg/) обмена нанопостами

и различные АИБ - [не единственное возможное место](https://github.com/username1565/nanoboard/issues/2) для паразитирования.

Одна из экстремальных идей - запихивание нанопостов в QR-коды и разбрасывание/расклеивание этих QR-кодов в местах,

где их будет удобно считывать (и, собственно, наклеивать/разбрасывать).

Так Наноборда сможет продолжить существование даже без Интернета, например, в метро.

Разумеется, вариант почти фантастический, но пример демонстрирует гибкость и ту самую неубиваемость Наноборды.

Ведь используя один лишь смартфон, с приложением в виде сканера qr-кодов,

даже в метро можно декодировать из qr-кода - коротенькую magnet-ссылку,

вытащить по ней - пак пикч из облачного торрента,

распарсить эти picture в бункере, на нейрокомпьютере, и залезть в нанабороду.

Алсо, может быть привинчена - IRC-координация (а это технология, она неубиваема, ага),

[Tox](https://tox.chat/download.html) (а там - [ассиметричная криптография](https://username1565.github.io/js-nacl/generate_keypair_and_crypt.html) на библиотеке [NaCl](https://username1565.github.io/js-nacl/)),

p2p (а там вообще - тотальная децентрализация пиринговых сетей же), 

и многое другое... 

## nanodb.exe-source from nanoboard-restore.zip

nanodb.exe-source from https://github.com/nanoboard/nanoboard/releases/download/0/nanoboard-restore.zip
nanoboard-restore\nanodb.exe-source
Version: Client 3.0
Language: C#

Q: How to compile nanodb.exe?
A:
1. Win+R
2. cmd, enter.
3. disk:\Path_to_.NET_Framework\MSBuild.exe /property:Configuration=Release nanodb.csproj

Q: Is nanodb.exe portable?
A: No, this working with files from restore-archive.

## Updates

See updates here: https://github.com/username1565/nanoboard/releases/

Commits: here: https://github.com/username1565/nanoboard/network


Changes.txt
Code commented. See source code.
Compiled version of all project - is the file "nanodb.exe"
See changes in other commits - here: https://github.com/username1565/nanoboard/network
		
nanodb.exe - this is a compiled binary file, as result of compilation this source code.
This is available in releases, here is an only source code of this, not a compiled exe.
		
BEGIN CHANGELOG:
__________________________________________________________________________________________________________________________________________________________________
(06.03.2021)	commit	(latest, unknown hash)	96 files changed with 46,571 additions and 46,464 deletions.
	Add captcha-generator.
		\nanodb.exe-source\Captcha\ByteStringExt.cs
			Add parse_number there.
		
		\nanodb.exe-source\Captcha\Captcha.cs
			Lines 19 - 23	:	Add usings for captcha-generator.
			Lines 48 - 55	:	Define "private sha" as null, and add "public static SHA256 _sha_public" to let get hash of captcha-pack.
			Lines 73 - 76	:	Define "private _packFile", and add "pubblic _packFile_public" to let get captcha-pack filename.
			Lines 277 - 280	:	Add this public methods to get values for current captcha.
			Line 413		:	Add comment.
			Lines 419 - 420	:	Add comments.
			Lines 423 - 975	:	Add code from https://github.com/username1565/nanoboard_Captcha_Pack_Generator/blob/8825757e25c183c0cf599d7486d10ec17c5b9437/src/Captcha_Pack_Generator.cs
								And comment unused code from standalone captcha-generator.

		\nanodb.exe-source\Program.cs
			Change EOL from LF to CRLF
			Lines 73 - 77	:	Add case to run captcha-generator by running nanodb with command-line argument "CaptchaGen".
		
		\Changes.txt - update changes in this file.
__________________________________________________________________________________________________________________________________________________________________
(27.02.2021)	commit	afad97384609b19774a7364407bed8bf32b8fff5	96 files changed with 46,571 additions and 46,464 deletions.
	Change EOL from LF to CRLF
		
Change "\n" to "\r\n" in compile.bat, for generated file "version.txt".
	\nanodb.exe-source\compile.bat
		
Add this changes, in this file - Changes.txt
	\Changes.txt
	
Change End-Of-Line, from LF to CRLF for 94 files.
	\nanodb.exe-source\Captcha\BitmapConvert.cs
	\nanodb.exe-source\Captcha\ByteStringExt.cs
	\nanodb.exe-source\Captcha\Captcha.cs
	\nanodb.exe-source\Captcha\PostSignatureExt.cs
	\nanodb.exe-source\Database\PostDb.cs
	\nanodb.exe-source\PngTransport\Aggregator.cs
	\nanodb.exe-source\PngTransport\CurlWebClient.cs
	\nanodb.exe-source\PngTransport\NBPack.cs
	\nanodb.exe-source\PngTransport\Post.cs
	\nanodb.exe-source\PngTransport\Posts.cs
	\nanodb.exe-source\PngTransport\WebClientX.cs
	\nanodb.exe-source\Server\ActionHandler.cs
	\nanodb.exe-source\Server\CaptchaTracker.cs
	\nanodb.exe-source\Server\Configurator.cs
	\nanodb.exe-source\Server\DbApiHandler.cs
	\nanodb.exe-source\Server\ErrorHandler.cs
	\nanodb.exe-source\Server\FileHandler.cs
	\nanodb.exe-source\Server\GetCaptchaImageHandler.cs
	\nanodb.exe-source\Server\GetCaptchaTokenHandler.cs
	\nanodb.exe-source\Server\HttpConnection.cs
	\nanodb.exe-source\Server\HttpRequest.cs
	\nanodb.exe-source\Server\HttpResponse.cs
	\nanodb.exe-source\Server\HttpServer.cs
	\nanodb.exe-source\Server\HttpServerBuilder.cs
	\nanodb.exe-source\Server\IRequestHandler.cs
	\nanodb.exe-source\Server\MimeType.cs
	\nanodb.exe-source\Server\NotificationHandler.cs
	\nanodb.exe-source\Server\SolveCaptchaAndAddPostHandler.cs
	\nanodb.exe-source\Server\StatusCode.cs
	\nanodb.exe-source\Server\StylesHandler.cs
	\nanodb.exe-source\Server\TcpServer.cs
	\nanodb.exe-source\nanodb.userprefs
	\pages\download_as_binary.html
	\pages\draw_image.html
	\pages\generate_date_for_version.html
	\pages\generate_HTTP(S)_links.html
	\pages\index.html
	\pages\Open_container.html
	\pages\params.html
	\pages\test_captcha.html
	\pages\TEST_download-upload-posts.html
	\pages\version.txt
	\pages_lite\download_as_binary.html
	\pages_lite\draw_image.html
	\pages_lite\generate_HTTP(S)_links.html
	\pages_lite\index.html
	\pages_lite\Open_container.html
	\pages_lite\params.html
	\pages_lite\TEST_download-upload-posts.html
	\pages_lite\version.txt
\scripts\api.js
	\scripts\checkver.js
	\scripts\img2base64.js
	\scripts\init.js
	\scripts\jquery-1.12.0.js
\scripts\last.js
\scripts\nanoclient.js
	\scripts\notif.js
	\scripts\on_add.js
	\scripts\params.js
\scripts\post.js
\scripts\reply.js
	\scripts\retranslate.js
\scripts\search.js
	\scripts\sha256.js
	\scripts_lite\api.js
	\scripts_lite\base64.js
	\scripts_lite\bootstrap.min.js
	\scripts_lite\checkver.js
	\scripts_lite\filesaver.min.js
	\scripts_lite\img2base64.js
	\scripts_lite\init.js
	\scripts_lite\jquery-1.12.0.js
	\scripts_lite\jquery.min.js
	\scripts_lite\last.js
	\scripts_lite\nanoclient.js
	\scripts_lite\notif.js
	\scripts_lite\on_add.js
	\scripts_lite\params.js
	\scripts_lite\post.js
	\scripts_lite\reply.js
	\scripts_lite\retranslate.js
	\scripts_lite\search.js
	\scripts_lite\sha256.js
	\scripts_lite\ui-functions.js
	\styles\skins\dark\nano.css
	\styles\skins\futaba\bootstrap.min.css
\styles\skins\futaba\nano.css
	\styles\skins\futaba\notif.css
	\styles\skins\photon\nano.css
	\styles\nano.css
	\externalIPservices.txt
		\proxy.txt
	\README.md

That's all.
__________________________________________________________________________________________________________________________________________________________________
(25.02.2021)	commit	338c00e1b3f4f7da22b906ad151fe313c078be4f 1 files changed with 53 additions and 2 deletions.
	Update Changes.txt

	/Changes.txt
		Update changes in /Changes.txt - add hashes of commits, and describe previous commits, in Changes.txt.
		Update additions and deletions, too.

scripts/post.js
function shortenHash(hash) {
  return hash.substring(0,4) + '..' + hash.substring(28,32);
}

function replaceAll(text, search, replace) {
  return text.split(search).join(replace);
}

function replaceAll_search(text, search, replace) {
	var result = text;
	if(
			search.indexOf("/")	===	-1	&&
			search.indexOf("[xmg")	===	-1
	){
		var regexp = new RegExp(search, 'gi');
		var match = text.match(regexp);
		if(match!=null && match.length==1 && match[0]!=regexp){
			replace = replace.split(regexp).join(match[0]);
			result = replaceAll(result, regexp, replace);
		}else if(match!=null && match.length>1){
			for(i=0; i<match.length; i++){
				//modify replace to current match
				replace = replace.split(regexp).join(match[i]);
				var t=0;   
				result =
					result.replace(regexp,
						function (match) {
							t++;
							return ((t-1) === i) ? replace : match;
						}
					)
				;
			}
		}
	}else{
		result = text.split(search).join(replace);
	}
  return result;
}

function escapeTags(text) {
  return text
    .replace(/>/gim, '&gt;')
    .replace(/</gim, '&lt;');
}

function detectImages(text) {
	var prefix = 'data:image/jpeg;base64,';
	//var matches = text.match(/\[(i|x)mg=[A-Za-z0-9+\/=]{4,64512}\]/g);	//images from karasiq nanoboard don't display as images, with limit.
  
	var regex = /\[(i|x)mg=[A-Za-z0-9+\/=]{4,}\]/g	;
	var result, indexes = [];
	while ( (result = regex.exec(text)) ) {
		indexes.push(result.index);
	}
	//console.log('matches', matches, ', indexes', indexes);					//array with indexes	
	var matches = text.match(regex);										//Now - OK.
	if (matches != null) {
		for (var i = 0; i < matches.length; i++) {
			var value = matches[i].toString();
			value = value.substring(5);
			value = value.substring(0, value.length - 1);
		
			var isfname = text.substring(indexes[i], text.length);		//get second part of message, before next value
			var isfname = isfname.split("]")[1];							//split by ']\n[filename.ext]' -> ( '' + "]" + '\n[filename.ext' + "]" )
			var isfname = isfname.substring(2, isfname.length);				//substring '\n[' to get filename.ext
			var test = /\.([0-9a-z]+)(?:[\?#]|$)/i.test(isfname);		//test is filename.ext there
			//console.log("detectImages, isfname", isfname, '(isfname==\'\')', (isfname==''), 'test', test);	//show result
	  
			if(isfname!=='' && test==true){		//if not empty string and if [filename.ext] after image
				var file_name = '['+isfname+']';	//add "[" and "]" to 'filename.ext' 
				var download_link = '<a href="'+prefix+value+'" download="'+isfname+'">['+isfname+']</a>';
				text = replaceAll(text, file_name, download_link);	//replace this to link for download the file.
				//console.log('<a href="'+prefix+value+'" download="'+isfname+'">['+isfname+']</a>'); //show this link
			}
	  
			value = '<img src="' + prefix + value + '" />';
			text = replaceAll(text, matches[i], value);
		}
	}
	//console.log('post.js: applyformatting, detectImages', '- img replaced to tag img');
	return text;
}

function addPlace(place,uuid) {
  place = Base64.decode(place);
  console.log('add:' + place);
  $.get('../api/paramget/places')
    .done(function(arr){
      arr = arr.split('\n');
      var wasAdded = arr.indexOf(place) != -1;
      if (wasAdded) {
			var elements = document.querySelectorAll('[id="'+uuid+'"]');
			for(i=0; i<elements.length; i++){
				$(elements[i]).text('added');	//add added for all elements
			}
//        $(document.getElementById(uuid)).text('added');	//old code
        pushNotification('Was added already.');
        return;
      }
      arr.push(place);
      $.post('../api/paramset/places', arr.join('\n'))
        .done(function(){
			var elements = document.querySelectorAll('[id="'+uuid+'"]');
			for(i=0; i<elements.length; i++){
				$(elements[i]).text('added');	//add added for all elements
			}
//        $(document.getElementById(uuid)).text('added');				//old code.
          pushNotification('Added: ' + place);
			updatePlacesBar_once();
        });
    });
}

function delPlace(place,uuid) {
  place = Base64.decode(place);
  console.log('del:' + place);
  $.get('../api/paramget/places')
    .done(function(arr){
      arr = arr.split('\n');
      var wasAdded = arr.indexOf(place) != -1;
      if (!wasAdded) {
			var elements = document.querySelectorAll('[id="'+uuid+'"]');	//get elements with the same id
			for(i=0; i<elements.length; i++){
				$(elements[i]).text('');									//remove "added" for all
			}
		//	$(document.getElementById(uuid)).text('');			//old code
        pushNotification('Not present or already deleted.');
        return;
      }
      var arr2 = [];
      for (var i = 0; i < arr.length; i++) {
        if (arr[i] != place) {
          arr2.push(arr[i]);
        }
      }
      arr = arr2;
      $.post('../api/paramset/places', arr.join('\n'))
        .done(function(){
			//console.log('delete: uuid: ', uuid);

			var elements = document.querySelectorAll('[id="'+uuid+'"]');	//get elements with the same id
			for(i=0; i<elements.length; i++){
				$(elements[i]).text('');						//remove "added" for all
			}
			
//          $(document.getElementById(uuid)).text('');		//old code
			pushNotification('Deleted: ' + place);
			updatePlacesBar_once();
        });
    });
}

function generateUUID(){
    var d = new Date().getTime();
    if(window.performance && typeof window.performance.now === "function"){
        d += performance.now(); //use high-precision timer if available
    }
    var uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
        var r = (d + Math.random()*16)%16 | 0;
        d = Math.floor(d/16);
        return (c=='x' ? r : (r&0x3|0x8)).toString(16);
    });
    return uuid;
}

function detectPlacesCommands(obj) {
  var text = obj.text();
  var html = obj.html();
  var incl = ''.includes == undefined ? function(x,y) { return x.contains(y); } : function(x,y) { return x.includes(y); };
  var matches = text.match(/(ADD|DEL(ETE|))[\s]*https?:\/\/[a-z%&\?\-=_\.0-9\/:#]+/g);
  if (matches == null) return;
  $.get('../api/paramget/places')
    .done(function(arr){
      arr = arr.split('\n');
      for (var i = 0; i < matches.length; i++) {
        var value = matches[i].toString();
        //console.log(value);
        value = value.substring(value.indexOf('http'));
        var wasAdded = arr.indexOf(value) != -1;
        
		//var uuid = generateUUID();								//old code using random UUID for each link
		var uuid = Sha256.hash(value);//.substring(0, 8);			//now using sha256 hash there, to add/remove "added" for all links in each post.
		
        html = replaceAll(html, value+'</a>', value +
          '</a>&nbsp;<a href=javascript:addPlace("'+Base64.encode(value)+'","'+uuid+'")><sup>[+]</sup></a>'+
          '<i><sup id="'+uuid+'">' + (wasAdded ? 'added' : '') + '</sup></i>' +
          '<a href=javascript:delPlace("'+Base64.encode(value)+'","'+uuid+'")><sup>[-]</sup></a>');
      }
      obj.html(html);
    });
}

//function to get link from text (+hash)
var get_link = function (value, post){
	return '<a target="_blank" href="'+value+((typeof post!=='undefined')?post:'')+'">'+value+'</a>';
}

function detectURLs(text) {
//  var matches = text.match(/https?:\/\/[A-Za-z%&\?\-=_\.0-9\/:#]+/g);	//old code not matching cyrillic symbols. "https://wiki.1chan.ca/Наноборда"

	//New regexp
	var matches = text.match(/(https?|ftps?|mailto|gopher|tox|irc|skype|magnet):?:\/\/(-\.)?[^\s\/?\.#<>]?[А-яЁёA-z0-9%~&()@\?\!\$\'\*\+\,\;\-=_\.\/:#]+/g);
	//cyrillic symbols in url - OK
	//more protocols - OK
	//dot " ", ".", "!", "?", ":", in the end of sentence not skiped. Can be "The link is http://example.org/index.html!!!" -> "http://example.org/index.html"
	//")" not skip when "(" not found in url earlier. Can be: "https://example.com/my(text).html))))))))" -> "https://example.com/my(text).html"
	
	//console.log('post.js - detectURLs(), mathes: ', matches);

  var you_re=new RegExp(".*youtube\.com.*")
  if (matches != null) {
    for (var i = 0; i < matches.length; i++) {
      var value = matches[i].toString();
	  
		//console.log('post.js detectURLs(), before replacement: matches['+i+']', value);
		var addition_string = '';
		if(
				(value.indexOf(')')	!==	-1)			//if closing bracket found in URL
			&&	(value.indexOf('(')	===	-1)			//and if opened bracket not found in URL
		){
			addition_string = value.substring(value.indexOf(')'), value.length);
			value = value.substring(0, value.indexOf(')'));	//leave URL before closed bracket.
		}
		//skip "]" in the end of link.
		if(
				(value.indexOf('[')	!==	-1)			//if closing bracket found in URL
			&&	(value.indexOf(']')	===	-1)			//and if opened bracket not found in URL
		){
			addition_string = value.substring(value.indexOf('['), value.length);
			value = value.substring(0, value.indexOf('['));	//leave URL before closed bracket.
		}
		
		if(
				(value.indexOf('(')	!==	-1)
			&&	(value.indexOf(')')	===	-1)
		){
			addition_string = value.substring(value.indexOf('('), value.length);
			value = value.substring(0, value.indexOf('('));	//leave URL before closed bracket.			
		}
		
		if(value[value.length-1] === '.'){
			addition_string = '.';
			value = value.substring(0, value.length-1);			 //remove dot in the end, if exist
		}
		
		//console.log('post.js detectURLs(), after replacement: matches['+i+']', value);
		//console.log('addition_string', addition_string);
	  
	  
      if (you_re.test(value))
      {
        value ='<a class="vd-vid" href="'+value+'"><span class="glyphicon glyphicon-play" aria-hidden="true"></span>'+value+'</a>';
        text = replaceAll(text, matches[i], value);
      }
      else
      {
		//current element - to link
        //value = '<a target=_blank href="'+value+'">'+value+'</a>';
        //replace in text current element to link
        //text = replaceAll(text, matches[i], value);
				//in this case, substrings replaced in replased strings.
			
			//do not do the double replace...
			//value = matches[i].toString();									//value for finding and replace. (defined earlier)
			var link = get_link(value)+addition_string;		//the link from value to insert
			var start_index = 0;												//increment start_index to search substrings, from this start_index
			
			for(j=0; j<matches.length; j++){									//for each url in array
				var next_url_index = text.indexOf(matches[j], start_index);		// search this url in text from start_index.
				if(next_url_index==-1){											//if this this url not found
					continue;													//continue find next url
				}																//else...
																			//continue the code...
				start_index = next_url_index; 									//else, if found, make this as start index.
				var maybe_href = text.substring(start_index-6, start_index);	//copy previous 6 symbols
				if(maybe_href==='href="'){										//if previous 6 symbols === 'href="' - do not replace text_link to link, this already replaced.
					var next_quote = text.indexOf('"', start_index);				//find next quote index.
					var url_length = next_quote-start_index;						//get length of URL
					start_index += url_length*2+6;									//move start_index 'url_length">url_length</a>'.length = url_length*2+6
				}else{															//if not href - then replace.
					var maybe_post = text.substring(start_index+matches[i].length+59, start_index+matches[i].length+59+32);
					if(
							typeof maybe_post !== 'undefined'
						&& 	(/[A-Fa-f0-9]{32}/g.test(maybe_post))
					){
						link = get_link(value, maybe_post)+addition_string;		//using post number inside the link, but leave link text.
					}
					text = 
							text.substring(0, start_index)									//substring by start_index
							+link															//insert link with link to first part
							+text.substring(start_index+matches[i].length, text.length);	//add second part
					break;
				}
			}
      }
    }
  }
  return text;
}

function detectThreadLinks(text) {
var matches = text.match(/&gt;&gt;[a-f0-9]{32}/g);
  if (matches != null) {
    for (var i = 0; i < matches.length; i++) {
      var value = matches[i].toString();
      value = value.substring(8, value.length);
      value = '<a href="javascript:void(0);" onclick=_depth=2;loadThread("'+value+'") title="Click to open post/thread/category...">&gt;&gt;' + value + '</a>';
      text = replaceAll(text, matches[i], value);
    }
  }
  return text;
}

function applyFormatting(text) {
	if(text.trim()===''){	//if empty post, from notif
		return false;			//do not do nothing...
	}
  text = text.replace(/&gt;(.*)/gi, "<gr>&gt;$1</gr>")
  text = text.replace(/\[sign=[a-f0-9]{128}\]/gim, '');
  text = text.replace(/\[pow=[a-f0-9]{256}\]/gim, '');
  text = text.replace(/\[sp(oiler|)\]/gim, '[x]');
  text = text.replace(/\[\/sp(oiler|)\]/gim, '[/x]');
  text = text.replace(/\[quote\]/gim, '<gr>&gt;');
  text = text.replace(/\[\/quote\]/gim, '</gr>');
  text = text.replace(/\[code\]/gim, '<pre style="display: inline-block; border: 0px; padding: 0px; margin: 0px;">');
  text = text.replace(/\[\/code\]/gim, '</pre>');
  text = text.replace(/\[sup\]/gim, '<sup>');
  text = text.replace(/\[\/sup\]/gim, '</sup>');
  text = text.replace(/\[sub\]/gim, '<sub>');
  text = text.replace(/\[\/sub\]/gim, '</sub>');
  text = text.replace(/\[o\]/gim, '<span style=text-decoration:overline>');
  text = text.replace(/\[\/o\]/gim, '</span>');

  var tags = 'biusxg';
  for (var x = 0; x < tags.length; x++) {
    var ch = tags.charAt(x);
    text = text
      .replace(new RegExp("\\[" + ch + "\\]", 'gim'), '<' + ch + '>')
      .replace(new RegExp("\\[/" + ch + "\\]", 'gim'), '</' + ch + '>');
  }
  text = detect_files(text);
  text = detectImages(text);
  text = replaceAll(text, 'data:image/jpeg;base64,iVBORw0K', 'data:image/png;base64,iVBORw0K');	//replace to jpeg to PNG if PNG signature found in base64
  text = detectThreadLinks(text);
  text = replaceAll(text, '\n', '<br/>');
  text = replaceAll(text, '  ', '&nbsp; ');
  if (_detectURLs == 'true') text = detectURLs(text);
  return text
    .replace(/<x>/gim, '<sp>')
    .replace(/<\/x>/gim, '</sp>');
}

(function($) {
  $.fn.extend({
    addTemporaryClass: function(className, duration) {
      var elements = this;
      setTimeout(function() {
        elements.removeClass(className);
      }, duration);
      return this.each(function() {
        $(this).addClass(className);
      });
    }
  });
})(jQuery);
